lib/nerve/reporter/zookeeper.rb

```diff
@@ -145,33 +159,114 @@ def encode_child_name(service)
     end
 
     def zk_delete
-      if @full_key
+      node_path = @full_key_mu.synchronize { @full_key }
+
+      if node_path
         statsd.time('nerve.reporter.zk.delete.elapsed_time', tags: ["zk_cluster:#{@zk_cluster}"]) do
-          @zk.delete(@full_key, :ignore => :no_node)
+          @zk.delete(node_path, :ignore => :no_node)
         end
-        @full_key = nil
+
+        @full_key_mu.synchronize {
```

Comment:
maybe create wrapper function for getting and setting with mutex?

Reply:
Sounds good! I'll separate this out into a `AtomicVariable` class that can be re-used.